### PR TITLE
Does `spark_udf` work with spark connect?

### PR DESCRIPTION
**Pull request by harupy**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

(Please fill in changes proposed in this fix)

## How is this patch tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: example.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+from pyspark.sql import SparkSession
+import mlflow
+import mlflow
+from mlflow.models import infer_signature
+from sklearn import datasets
+from sklearn.neighbors import KNeighborsClassifier
+from pyspark.sql import SparkSession
+
+spark = SparkSession.builder.getOrCreate()
```
Comment (harupy) [inline example.py:9]: `with SparkSession.builder.getOrCreate()` didn't work.

File: NOTES.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+Traceback (most recent call last):
+  File "example.py", line 23, in <module>
+    pyfunc_udf = mlflow.pyfunc.spark_udf(spark, model_info.model_uri, env_manager="conda")
+  File "/Users/harutakakawamura/Desktop/repositories/mlflow/mlflow/pyfunc/__init__.py", line 973, in spark_udf
+    result_type = _parse_datatype_string(result_type)
+  File "/Users/harutakakawamura/.pyenv/versions/miniconda3-4.7.12/envs/mlflow-dev-env/lib/python3.8/site-packages/pyspark/sql/types.py", line 1212, in _parse_datatype_string
+    sc = get_active_spark_context()
+  File "/Users/harutakakawamura/.pyenv/versions/miniconda3-4.7.12/envs/mlflow-dev-env/lib/python3.8/site-packages/pyspark/sql/utils.py", line 202, in get_active_spark_context
+    raise RuntimeError("SparkContext or SparkSession should be created first.")
+RuntimeError: SparkContext or SparkSession should be created first.
```
Comment (harupy) [inline NOTES.md:27]: @WeichenXu123 Looks like `get_active_spark_context` is not ready for Spark Connect.

File: NOTES.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
+## Set up
+
+```sh
+wget https://dlcdn.apache.org/spark/spark-3.4.1/spark-3.4.1-bin-hadoop3.tgz
+mkdir spark-bin
+tar -xvf spark-3.4.1-bin-hadoop3.tgz --directory spark-bin
+./spark-bin/spark-3.4.1-bin-hadoop3/sbin/start-connect-server.sh --packages org.apache.spark:spark-connect_2.12:3.4.0
```
Comment (harupy) [inline NOTES.md:7]: Is this the only way to launch a spark connect server?

File: NOTES.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+```
+
+```
+/Users/harutakakawamura/.pyenv/versions/miniconda3-4.7.12/envs/mlflow-dev-env/lib/python3.8/site-packages/_distutils_hack/__init__.py:33: UserWarning: Setuptools is replacing distutils.
+  warnings.warn("Setuptools is replacing distutils.")
+Traceback (most recent call last):
+  File "example.py", line 23, in <module>
+    pyfunc_udf = mlflow.pyfunc.spark_udf(spark, model_info.model_uri, env_manager="conda")
+  File "/Users/harutakakawamura/Desktop/repositories/mlflow/mlflow/pyfunc/__init__.py", line 973, in spark_udf
+    result_type = _parse_datatype_string(result_type)
```
Comment (harupy) [inline NOTES.md:22]: This is unrelated, but we should never use a private function like `_parse_datatype_string`.

File: NOTES.md (hunk 2 above)
Comment (WeichenXu123) [reply NOTES.md:27]: yes, but we can use `pyspark.sql.connect.session._active_spark_session` instead.

File: NOTES.md (hunk 3 above)
Comment (WeichenXu123) [reply NOTES.md:7]: for local test, we can use `bin/pyspark --remote local` to start connect server locally and start a spark python REPL at the same time

File: mlflow/pyfunc/__init__.py [hunk 5]
```diff
@@ -1067,7 +1067,8 @@ def _predict_row_batch(predict_fn, args):
                 names = [str(i) for i in range(len(args))]
             else:
                 names = input_schema.input_names()
-                required_names = input_schema.required_input_names()
+                # required_names = input_schema.required_input_names()
+                required_names = names
```
Comment (harupy) [inline mlflow/pyfunc/__init__.py:1071]: I used MLR 13.1. MLflow 2.3.1 is installed there, but mlflow in my machine is 2.4.2.dev0. 2.3.1 doesn't have the `required_input_names` attribute.

File: mlflow/pyfunc/__init__.py [hunk 6]
```diff
@@ -1272,7 +1273,11 @@ def batch_predict_fn(pdf):
             if should_use_spark_to_broadcast_file:
                 loaded_model, _ = SparkModelCache.get_or_load(archive_path)
             else:
-                loaded_model = mlflow.pyfunc.load_model(local_model_path)
+                with tempfile.TemporaryDirectory() as tmpdir:
+                    local_model_path = _download_artifact_from_uri(
+                        artifact_uri=model_uri, output_path=tmpdir
+                    )
+                    loaded_model = mlflow.pyfunc.load_model(local_model_path)
```
Comment (harupy) [inline mlflow/pyfunc/__init__.py:1280]: Another hack.

File: NOTES.md [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+./spark-bin/spark-3.4.1-bin-hadoop3/sbin/start-connect-server.sh --packages org.apache.spark:spark-connect_2.12:3.4.0
+export SPARK_REMOTE="sc://localhost"
+```
+
+```sh
+python example.py
+```
+
+```
+/Users/harutakakawamura/.pyenv/versions/miniconda3-4.7.12/envs/mlflow-dev-env/lib/python3.8/site-packages/_distutils_hack/__init__.py:33: UserWarning: Setuptools is replacing distutils.
```
Comment (harupy) [inline NOTES.md:17]: It turns out the cause of this error was a spark connect server wasn't running.

**Comment by harupy** [comment]
@dbczumar I requested a review but a review is not needed, just wanted to share what I did and the issues I found.

File: mlflow/pyfunc/__init__.py (hunk 6 above)
Comment (harupy) [reply mlflow/pyfunc/__init__.py:1280]: This worked.

Comment (harupy) [reply mlflow/pyfunc/__init__.py:1280]: - Can UDF write to NFS? -> No, need a new API? let's ask spark connect team.

**Pull request closed without merging**